examples/js/nodes/materials/MeshStandardNodeMaterial.js

```diff
@@ -6,21 +6,29 @@ import { MeshStandardNode } from './nodes/MeshStandardNode.js';
 import { NodeMaterial } from './NodeMaterial.js';
 import { NodeUtils } from '../core/NodeUtils.js';
 
-function MeshStandardNodeMaterial() {
+function MeshStandardNodeMaterial( parameters ) {
```

Comment:
I'm not sure about this. I never liked passing parameters as an object, it's error-prone and limited.
We have a chance to move away from it here.

Reply:
If you'd like to move away from parameter objects this does seem like a good chance. But then I'm not sure I understand the difference between `MeshStandardNodeMaterial` and `StandardNodeMaterial` — I thought the goal of this class was to be a backward-compatible alternative to `MeshStandardMaterial`, whereas `StandardNodeMaterial` would provide a new, fully node-based syntax?

Reply:
@sunag is that correct, or what would be the goal with the two material classes here? Put another way, is this what we want?

- MeshStandardMaterial (current)
- MeshStandardNodeMaterial (node-based, fully backward-compatible)
- StandardNodeMaterial (node-based, new API, new features)

I’m assuming here that MeshStandardMaterial would eventually be replaced by MeshStandardNodeMaterial. If that’s not the goal, or if we want the existing system to exist alongside node-based materials indefinitely, then perhaps we don’t need a backward-compatible MeshStandardNodeMaterial?

Reply:
I think the plan was to make `MeshStandardMaterial` node-based to keep backwards-compatibility.

Maybe `MeshStandardMaterial` could extend `MeshStandardNodeMaterial`.

Reply:
> I never liked passing parameters as an object

Personally, I much prefer `parameters` as an object to
```javascript
Texture( image, mapping, wrapS, wrapT, magFilter, minFilter, format, type, anisotropy, encoding )
```
where the order matters. Very annoying.

Reply:
> I think the plan was to make MeshStandardMaterial node-based to keep backwards-compatibility.

That sounds like a fine plan to me. But once that happens, there's probably no reason to have three classes involved. We could just delete MeshStandardMaterial, rename MeshStandardNodeMaterial as MeshStandardMaterial, and if everything is backward-compatible there is no user-facing change. Users should only know about MeshStandardMaterial and the new NodeStandardMaterial.

Assuming we're not quite ready to make that change yet, having MeshStandardNodeMaterial _be_ backward-compatible even before it replaces MeshStandardMaterial seems nice, because it makes it easier for users to drop it into existing scenes (or loaders, like this PR) test the node material system out, and give early feedback.

In that case maybe MeshStandardNodeMaterial should be renamed to something more obviously temporary, like MeshStandardMaterial2?

> I never liked passing parameters as an object
>
> > Personally, I much prefer parameters as an object to [ordered parameters]

+1 that ordered parameters are frustrating, @mrdoob did you mean that, or having no constructor arguments at all here?

Reply:
I agree that passing parameters like `Texture` currently does is also annoying.

I prefer this kind of API:

```js
new THREE.Texture( image ).setMapping( mapping ).setFormat( format );
```

Reply:
> We could just delete MeshStandardMaterial, rename MeshStandardNodeMaterial as MeshStandardMaterial, and if everything is backward-compatible there is no user-facing change. Users should only know about MeshStandardMaterial and the new NodeStandardMaterial.

I think that's too much of a big change. Too many tutorials/articles that will get obsolete.

I think it's easier to think about `*NodeMaterial` as having a different API. If we bring parameters to the new API we're carrying over the problem and it'll be harder to maintain.

Reply:
Hm... I don't think that we should introduce _two_ new standard materials (`MeshStandardNodeMaterial`, `StandardNodeMaterial`) that both have different, breaking APIs for the same features. For the new API, `StandardNodeMaterial` seems like the place to start — we can leave parameter support out there. If `MeshStandardNodeMaterial` will not have the same API as `MeshStandardMaterial`, I think we should just delete it (`MeshStandardNodeMaterial`) to prevent confusion, unless it has some other purpose I've missed.

Reply:
> I never liked passing parameters as an object, it's error-prone and limited.

Can you explain why this:

```js 
const mat = new MeshStandardMaterial( { color: oxff0000, roughness: 0.25, transparency: true, map, alphaMap, normalMap } ) ;
```

is more error prone or limited than

```js
const mat = new MeshStandardMaterial().setColor( 0xff0000).setRoughness( 0.25 ).setTransparent( true ).setMap( map ).setAlphaMap( alphaMap ).setNormalMap( normalMap );
```

The second approach seems much more verbose to me, with no advantage in clarity - especially since this would be a breaking change.

Reply:
> But the problem of the last one is that the user could make a typo like texture.fromat and javascript won't complain. Unless we start using Object.freeze().

If that is the problem I think this could be fixed with `hasOwnProperty` declaring all `properties` before of `setValues` even `undefined` variables:

Example (1)
```javascript
var params = { format : undefined };

// inside while
var prop = "fromat", val = "any value";

if ( params.hasOwnProperty( prop  ) ) params[prop] = val;
else console.warn(prop, "property does not exist or has become obsolete.");
```

Example( 2)
```javascript
NodeMaterial.prototype.setValues = function( values ) {

	if ( values === undefined ) return;

	for ( var key in values ) {

		if ( this.hasOwnProperty( key ) ) {

			this[ key ] = values[ key ];

		} else {

			console.warn(key, "property does not exist or has become obsolete.");

		}

	}

};
```

> Personally, I much prefer parameters as an object to [ordered parameters]

+1 I think this could come from the an evolution of compiler too like `interface` of `typescript`. For me, besides being practical to define parameters it techniques is good to avoid changes if a property is to be depreciate in relation to ordered parameters, in relation at `setParamA(a).setParamB(b)` could help to reduce the code.

Another third thing could be to create a classe just for this.
```javasacript
new THREE.SetValues( this, params );
// or
THREE.SetValues( this, params )
```